cpp/mrc/src/public/coroutines/task_container.cpp

```diff
@@ -133,6 +140,19 @@ auto TaskContainer::gc_internal() -> std::size_t
     return deleted;
 }
 
+void TaskContainer::start_next_task()
+{
+    auto pos = [this]() {
+        std::scoped_lock lk{m_mutex};
+        auto pos = m_next_tasks.front();
+        m_next_tasks.pop();
+        return pos;
+    }();
+
+    // Start executing from the cleanup task to schedule the user's task onto the thread pool.
+    pos->value().resume();
```

Comment:
I think this could be simplified by converting `m_mutex` to a `recursive_mutex`. Then you dont need to worry about deadlocks from the same thread.

Reply:
it's possible a different thread can execute the coroutine, so we need to make sure there's no lock at this point.